File: flink-api-scala-example/src/main/scala/com/ash/flink/tfm/IterateTest.scala
package com.ash.flink.tfm

import org.apache.flink.streaming.api.scala.{DataStream, StreamExecutionEnvironment}

object IterateTest {

  def main(args: Array[String]): Unit = {
    val env = StreamExecutionEnvironment.getExecutionEnvironment
    val socketDs: DataStream[String] = env.socketTextStream("localhost", 9999)

    import org.apache.flink.streaming.api.scala._
    val socketIntDs = socketDs.map(line => line.toInt)

    val result = socketIntDs.iterate(iteration => {
      //定义迭代体
      val minusOne: DataStream[Int] = iteration.map(v => {println("迭代体中value值为："+v);v - 1})
      //定义迭代条件，满足的继续迭代
      val stillGreaterThanZero: DataStream[Int] = minusOne.filter(_ > 0)
      //定义哪些数据最后进行输出
      val lessThanZero: DataStream[Int] = minusOne.filter(_ <= 0)
      //返回tuple，第一个参数是哪些数据流继续迭代，第二个参数是哪些数据流进行输出
      (stillGreaterThanZero, lessThanZero)
    })

    result.print()
    env.execute()


  }


}

File: flink-api-java-example/src/main/java/com/ash/flink/tfm/IterateTest.java
package com.ash.flink.tfm;

import org.apache.flink.api.common.functions.FilterFunction;
import org.apache.flink.api.common.functions.MapFunction;
import org.apache.flink.streaming.api.datastream.DataStreamSource;
import org.apache.flink.streaming.api.datastream.IterativeStream;
import org.apache.flink.streaming.api.datastream.SingleOutputStreamOperator;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;

public class IterateTest {

    public static void main(String[] args) throws Exception {
        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
        DataStreamSource<String> socketDs = env.socketTextStream("localhost", 9999);
        SingleOutputStreamOperator<Integer> convertInt = socketDs.map(new MapFunction<String, Integer>() {
            @Override
            public Integer map(String value) throws Exception {
                return Integer.parseInt(value);
            }
        });

        IterativeStream<Integer> iterateDs = convertInt.iterate();
        SingleOutputStreamOperator<Integer> iterOperator = iterateDs.map(new MapFunction<Integer, Integer>() {
            @Override
            public Integer map(Integer value) throws Exception {
                return value - 1;
            }
        });

        SingleOutputStreamOperator<Integer> continueOperator = iterOperator.filter(new FilterFunction<Integer>() {
            @Override
            public boolean filter(Integer value) throws Exception {
                return value > 0;
            }
        });


        iterateDs.closeWith(continueOperator);
        iterateDs.print();
        env.execute();


    }


}
